### PR TITLE
Refactored PulsarResources to avoid building metadata paths everywhere

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

The `PulsarResources` and its companion classes are all taking `String`s with paths for resources in the metadata store. This has lead to a proliferation of string composition to construct these paths from all over the code base. 

Instead, we should be taking the resource namings (eg: `TopicName`, `NamespaceName`, etc.) and hide the path in the metadata.

### Modifications

 * Changes all methods in `BaseResources` to be protected
 * Each subclass exposes resource specific public methods

### CODE REVIEW

**Comment by Anonymitaet** [comment]
Thanks for your contribution. For this PR, do we need to update docs?

(The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

**Comment by merlimat** [comment]
> Thanks for your contribution. For this PR, do we need to update docs?
 @Anonymitaet no need for docs changes here. Just labeled.

**Review by sijie — APPROVED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/ClusterResources.java [hunk 1]
```diff
@@ -48,5 +104,33 @@ public FailureDomainResources(MetadataStore store, Class<FailureDomainImpl> claz
...
+            return get(joinPath(BASE_CLUSTERS_PATH, clusterName, FAILURE_DOMAIN, domainName));
+        }
+
+        public void deleteFailureDomain(String clusterName, String domainName) throws MetadataStoreException {
+            delete(joinPath(BASE_CLUSTERS_PATH, clusterName, FAILURE_DOMAIN, domainName));
+        }
+
+        public void deleteFailureDomains(String clusterName) throws MetadataStoreException {
+            String failureDomainPath = joinPath(BASE_CLUSTERS_PATH, clusterName, FAILURE_DOMAIN);
+            for (String domain : getCache().getChildren(failureDomainPath).join()) {
```
Comment (eolivelli) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/ClusterResources.java:123]: Is it correct to use 'cache' here?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java [hunk 2]
```diff
@@ -36,17 +38,39 @@
         this.store = store;
     }
 
+    public CompletableFuture<List<String>> listPersistentTopicsAsync(NamespaceName ns) {
+        String path = "/managed-ledgers/" + ns + "/persistent";
```
Comment (eolivelli) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java:42]: Nit: use constant ?

**Review by eolivelli — APPROVED** [review]
Well done.
thIs is is very useful clean up.
Thanks for doing it

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java (hunk 2 above)
Comment (merlimat) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java:42]: 👍

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/ClusterResources.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/ClusterResources.java:123]: In this case it's redundant but correct. The exists and getChildren are cached directly in the MetadataStore implementation, the MetadataCache is for keeping a typed cache of the objects. 
The delete operation on the cache is just a convenience method that is just forwarded to the store implementation

Comment (eolivelli) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/ClusterResources.java:123]: Makes sense. Thanks

**Pull request merged**